example_buffer.py

```diff
@@ -151,63 +156,74 @@ def _rsync_dir(source_dir, dest_dir):
 
 
 def _determine_chunk_to_make(write_dir):
+    """
+    Returns the full path of the chunk to make (gs://...)
+    and a boolean, indicating whether we should wait for a new model
+    or if we're 'behind' and should just write out our current chunk immediately
+    True == write immediately.
+    """
     models = fsdb.get_models()
-    # Last model is N.  N+1 is training.  We should gather games for N+2.
+    # Last model is N.  N+1 (should be) training.  We should gather games for N+2.
+    chunk_to_make = os.path.join(write_dir, str(
+        models[-1][0] + 1) + '.tfrecord.zz')
+    if not tf.gfile.Exists(chunk_to_make):
+        # N+1 is missing.  Write it out ASAP
+        print("Making chunk ASAP:", chunk_to_make)
+        return chunk_to_make, True
     chunk_to_make = os.path.join(write_dir, str(
         models[-1][0] + 2) + '.tfrecord.zz')
     while tf.gfile.Exists(chunk_to_make):
         print("Chunk for next model ({}) already exists.  Sleeping.".format(chunk_to_make))
         time.sleep(5 * 60)
+        models = fsdb.get_models()
+        chunk_to_make = os.path.join(write_dir, str(
+            models[-1][0] + 2) + '.tfrecord.zz')
     print("Making chunk:", chunk_to_make)
 
-    return chunk_to_make
+    return chunk_to_make, False
 
 def fill_and_wait_time(bufsize=dual_net.EXAMPLES_PER_GENERATION,
                   write_dir=None,
                   threads=32,
-                  start_from=dt.datetime.utcnow()):
+                  start_from=None):
+    start_from = start_from or dt.datetime.utcnow()
     write_dir = write_dir or fsdb.golden_chunk_dir()
     buf = ExampleBuffer(bufsize)
-    chunk_to_make = _determine_chunk_to_make(write_dir)
+    chunk_to_make, fast_write = _determine_chunk_to_make(write_dir)
 
-    hours = _get_hour_dirs()
+    hours = fsdb.get_hour_dirs()
     with timer("Rsync"):
-        time_rsync(_earlier_of(dt.datetime.strptime(hours[-1], "%Y-%m-%d-%H/"), start_from))
+        time_rsync(min(dt.datetime.strptime(hours[-1], "%Y-%m-%d-%H/"), start_from))
         start_from = dt.datetime.utcnow()
 
-    hours = _get_hour_dirs()
+    hours = fsdb.get_hour_dirs()
+    files = (tf.gfile.Glob(os.path.join(LOCAL_DIR, d, "*.zz"))
```

Comment:
is this missing an itertools.chain? I think `files` here is a generator of lists of filepaths, not a generator of filepaths.